fvm/src/machine/manifest.rs

```diff
@@ -0,0 +1,129 @@
+use std::collections::{HashMap, HashSet};
+
+use anyhow::{anyhow, Context};
+use cid::Cid;
+use fvm_ipld_blockstore::Blockstore;
+use fvm_ipld_encoding::CborStore;
+
+const SINGLETON_ACTOR_NAMES: &[&str] = &[
+    "system",
+    "init",
+    "reward",
+    "cron",
+    "storagepower",
+    "storagemarket",
+    "verifiedregistry",
+];
+
+const ACCOUNT_ACTOR_NAME: &str = "account";
+const INIT_ACTOR_NAME: &str = "init";
+const SYSTEM_ACTOR_NAME: &str = "system";
+
+/// A mapping of builtin actor CIDs to their respective types.
+pub struct Manifest {
+    account_code: Cid,
+    system_code: Cid,
+    init_code: Cid,
+    singletons: HashSet<Cid>,
+
+    by_id: HashMap<u32, Cid>,
+    by_code: HashMap<Cid, u32>,
+}
+
+impl Manifest {
+    pub fn builtin_actor_codes(&self) -> impl Iterator<Item = &Cid> {
+        self.by_id.values()
+    }
+
+    /// Returns the code CID for a builtin actor, given the actor's ID.
+    pub fn code_by_id(&self, id: u32) -> Option<&Cid> {
+        self.by_id.get(&id)
+    }
+
+    /// Returns the code CID for the account actor.
+    pub fn get_account_code(&self) -> &Cid {
+        &self.account_code
+    }
+
+    /// Returns the code CID for the init actor.
+    pub fn get_init_code(&self) -> &Cid {
+        &self.init_code
+    }
+
+    /// Returns the code CID for the system actor.
+    pub fn get_system_code(&self) -> &Cid {
+        &self.system_code
+    }
+
+    /// Returns the the actor code's "id" if it's a builtin actor. Otherwise, returns 0.
+    pub fn id_by_code(&self, code: &Cid) -> u32 {
+        self.by_code.get(code).copied().unwrap_or(0)
+    }
+
+    /// Returns true id the passed code CID is the account actor.
+    pub fn is_account_actor(&self, cid: &Cid) -> bool {
+        &self.account_code == cid
+    }
+
+    /// Returns true id the passed code is a singleton actor.
+    pub fn is_singleton_actor(&self, cid: &Cid) -> bool {
+        self.singletons.contains(cid)
+    }
+
+    /// Load a manifest from the blockstore.
+    pub fn load<B: Blockstore>(bs: &B, root_cid: &Cid, ver: u32) -> anyhow::Result<Manifest> {
+        if ver != 1 {
+            return Err(anyhow!("unsupported manifest version {}", ver));
+        }
+
+        let vec: Vec<(String, Cid)> = match bs.get_cbor(root_cid)? {
+            Some(vec) => vec,
+            None => {
+                return Err(anyhow!("cannot find manifest root cid {}", root_cid));
+            }
+        };
+        Manifest::new(vec)
+    }
+
+    /// Construct a new manifest from actor name/cid tuples.
+    pub fn new(iter: impl IntoIterator<Item = (String, Cid)>) -> anyhow::Result<Self> {
+        let mut by_name = HashMap::new();
+        let mut by_id = HashMap::new();
+        let mut by_code = HashMap::new();
+        for ((name, code_cid), id) in iter.into_iter().zip(1u32..) {
+            by_id.insert(id, code_cid);
```

Comment:
shouldnt we sort those?

Reply:
We need to maintain ordering as they appear in the manifest. Basically, the actor bundle is the ultimate source of truth here.